julia_src/reopt_model.jl

```diff
@@ -1145,8 +1145,202 @@ function add_cost_function(m, p)
 		it is arbitrary where the min charge ends up (eg. could be in m[:TotalDemandCharges] or m[:MinChargeAdder]).
 		0.0001*m[:MinChargeAdder] is added back into LCC when writing to results.  =#
+
+	# Add climate costs
+	if p.Include_climate_in_objective # if user selects to include climate in objective
+		add_to_expression!(Costs, m[:Lifecycle_Emissions_Cost_CO2])
+	end
+
+	# Add Health costs (NOx, SO2, PM2.5)
+	if p.Include_health_in_objective
+		add_to_expression!(Costs, m[:Lifecycle_Emissions_Cost_Health])
+	end
+
+end
+
+# Renewable electricity calculation
+function add_re_elec_calcs(m,p)
+	if p.IncludeExportedREElecinTotal
+		include_exported_re_elec_in_total = 1
+	else
+		include_exported_re_elec_in_total = 0
+	end
+	m[:AnnualREEleckWh] = @expression(m,p.TimeStepScaling*
+		(sum(p.ProductionFactor[t,ts] * p.LevelizationFactor[t] * m[:dvRatedProduction][t,ts] * p.TechPercentRE[t] for t in p.ElectricTechs, ts in p.TimeStep) #total RE elec generation, excl steam turbine
+		- sum(m[:dvProductionToStorage][b,t,ts]*p.TechPercentRE[t]*(1-p.ChargeEfficiency[t,b]*p.DischargeEfficiency[b]) for t in p.ElectricTechs, b in p.ElecStorage, ts in p.TimeStep) #minus battery efficiency losses
+		- sum(m[:dvProductionToCurtail][t,ts]*p.TechPercentRE[t] for t in p.ElectricTechs, ts in p.TimeStep) # minus curtailment.
+		- (1-include_exported_re_elec_in_total)*sum(m[:dvProductionToGrid][t,u,ts]*p.TechPercentRE[t] for t in p.ElectricTechs,  u in p.ExportTiersByTech[t], ts in p.TimeStep) # minus exported RE, if RE accounting method = 0.
+		+ p.STElecOutToThermInRatio * sum(m[:dvThermalToSteamTurbine][tst,ts]*p.TechPercentRE[tst] for ts in p.TimeStep, tst in p.TechCanSupplySteamTurbine) )) # plus steam turbine generation (steam turbine p.TechPercentRE input = 0 because it is a decision variable dependent on fraction of steam generated by RE fuel
+		# TODO: Steam Turbine battery losses and curtailment (challenge is that equations seem non-linear)
+		# Note: if battery ends up being allowed to discharge to grid, need to make sure only RE that is being consumed onsite is counted so battery doesn't become a back door for RE to grid.
+end
+
+
+#Renewable electricity constraints
+function add_re_elec_constraints(m,p)
+	if !isnothing(p.MinAnnualPercentREElec)
+		@constraint(m, MinREElecCon, m[:AnnualREEleckWh] >= p.MinAnnualPercentREElec*p.TimeStepScaling*(sum(p.ElecLoad[ts] for ts in p.TimeStep)))
+	end
+	if !isnothing(p.MaxAnnualPercentREElec)
+		@constraint(m, MaxREElecCon, m[:AnnualREEleckWh] <= p.MaxAnnualPercentREElec*p.TimeStepScaling*(sum(p.ElecLoad[ts] for ts in p.TimeStep)))
+	end
+end
+
+#Renewable heat calculations
+function add_re_heat_calcs(m,p)
+	# Renewable heat
+	m[:AnnualREHeatMMBTU] = @expression(m,p.TimeStepScaling*
+		(sum(p.ProductionFactor[t,ts] * p.LevelizationFactor[t] * m[:dvThermalProduction][t,ts] * p.TechPercentRE[t] for t in p.HeatingTechs, ts in p.TimeStep) #total RE heat generation (excl steam turbine, GHP)
```

Comment:
ST is included in `HeatingTechs`, however, I think we do **not** need to include ST thermal production because that is always a subset of the heat **into** the ST that is supplied by one of the other `HeatingTechs`. So all of the emissions from the `HeatingTechs` (excluding ST) are accounted for. So I think we need to explicitly remove ST from the set of `HeatingTechs` in this summation.

GHP: I think we do **not** need to do anything with GHP necessarily if we don't need to track the GHP contribution to %RE (I don't think we do). GHP removes the fuel consumption (emissions) from the Boiler and replaces it with extra electricity consumption which can be served by on-site generation or the grid.

Reply:
Ok so if ST is in HeatingTechs, that means that right now we are counting ST thermal production twice. For the ST term below, instead of adding ST thermal production and subtracting the thermal production from techs feeding ST, we would just want to do the subtraction. "+ (p.STThermOutToThermInRatio-1) * sum(...)" becomes "- sum(...)"

Reply:
@hdunham don't we want to keep the `dvThermalProduction["SteamTurbine"]` in the first term, but then **subtract** out the full `sum(dvThermalToSteamTurbine)` because the latter is not actually serving the heat load, while the former is. This is counter to what I said above, so ignore that. So in the last term, we'd remove `(p.STThermOutToThermInRatio - 1)`.

Reply:
Yes, we are saying the same thing. Except that the + in front of that term needs to to become a -.

Reply:
I agree after re-reading your comment we are saying the same thing.

Reply:
Hi, I agree about the approach for SteamTurbines (essentially subtracting out "losses" using the STThermOutToThermInRatio, similar to how we handle storage losses). 

However, I was thinking that we should, if possible (I'm worried it's nonlinear, similar to the SteamTurbine RE elec calculation...) calculate the overall percent RE elec mix (calculated as total RE elec consumption divided by total consumption; though I'm not sure whether it would be okay to do total annual percent or keep it per timestep) and then apply that as essentially the p.TechPercentRE value for GHP (though obviously not using that variable itself)

Reply:
@kkrah  I don't think GHP should have a p.TechPercentRE because it's not a supply technology. I think of it more like a load. So I think we don't need to do anything with GHP. I'm happy to chat if you still think we need to explicitly handle GHP some way.

Reply:
I agree GHP kinda makes things weird, but I'm trying to follow how GHP is double-counted for in RE-elec and RE-heat. I don't see where GHP is counted in the RE-heat equation. It does subtract from the load that the Boiler would otherwise serve by burning fuel - is that where you are saying it is counted for in the RE-heat? Either way, since we are not doing heat RE-% constraints, I agree we can probably just document this for the RE-heat accounting.

On the electric side, GHP is adding to the net load (sometimes, if electrifying the heat adds more electricity than the reduction in electricity for cooling) because it's electrifying the otherwise Boiler (fuel) heating load. In that case, it will require more RE generation to hit the %-RE target. There's not really an issue here, just noting that it might be a non-intuitive result.

I also just realized a general issue of the basis the RE constraint: I see we are using `p.ElectricLoad` as the basis, but that is the "non-cooling electric load" so I think we need to add `p.CoolingLoad[ts] / p.ElectricChillerCOP` to that.

See [electric load additions to p.ElectricLoad (non-cooling electric load)](https://github.com/NREL/REopt_Lite_API/blob/d4f21f128e6ade5f925c15cdfbf013417011e862/julia_src/reopt_model.jl#L796) and 3 lines below that for AbsorptionChiller and GHP on the "load-side" of the load balance equation. I think we might need to note that AbsorptionChiller might mess up RE-% accounting too because it takes a potentially RE fuel and serves the cooling load which reduces the cooling electricity consumption from the ElectricChiller.